txscript/consensus.go

```diff
@@ -65,3 +66,278 @@ func ExtractCoinbaseNullData(pkScript []byte) ([]byte, error) {
+	// interpreted as a negative number.
+	if sLen > 1 && sig[sOffset] == 0x00 && sig[sOffset+1]&0x80 == 0 {
+		str := "malformed signature: S value has too much padding"
+		return scriptError(ErrSigTooMuchSPadding, str)
+	}
+
+	// Verify the S value is <= half the order of the curve.  This check is done
+	// because when it is higher, the complement modulo the order can be used
+	// instead which is a shorter encoding by 1 byte.
+	sValue := new(big.Int).SetBytes(sig[sOffset : sOffset+sLen])
+	if sValue.Cmp(halfOrder) > 0 {
+		return scriptError(ErrSigHighS, "signature is not canonical due to "+
+			"unnecessarily high S value")
+	}
+
+	return nil
+}
+
+// IsStrictSignatureEncoding returns false if the passed signature does not
+// adhere to the strict encoding requirements.
+func IsStrictSignatureEncoding(signature []byte) bool {
+	return CheckSignatureEncoding(signature) == nil
+}
+
+// CheckPubKeyEncoding returns an error if the passed public key does not
+// adhere to the strict encoding requirements.
+func CheckPubKeyEncoding(pubKey []byte) error {
+	if !isStrictPubKeyEncoding(pubKey) {
+		return scriptError(ErrPubKeyType, "unsupported public key type")
+	}
+	return nil
+}
+
+// CheckHashTypeEncoding returns whether or not the passed hashtype adheres to
+// the strict encoding requirements.
+func CheckHashTypeEncoding(hashType SigHashType) error {
+	sigHashType := hashType & ^SigHashAnyOneCanPay
+	if sigHashType < SigHashAll || sigHashType > SigHashSingle {
+		str := fmt.Sprintf("invalid hash type 0x%x", hashType)
+		return scriptError(ErrInvalidSigHashType, str)
+	}
+	return nil
+}
+
+// IsStrictCompressedPubKeyEncoding returns whether or not the passed public
+// key adheres to the strict compressed encoding requirements.
+func IsStrictCompressedPubKeyEncoding(pubKey []byte) bool {
+	if len(pubKey) == 33 && (pubKey[0] == 0x02 || pubKey[0] == 0x03) {
+		// Compressed
+		return true
+	}
+	return false
+}
+
+// IsStrictNullData returns whether or not the passed data is an OP_RETURN
+// followed by specified length data push. It explicitly verifies that the
+// opcode is identical to the expected length. This function does not support
+// checks for data > 75 bytes.
+func IsStrictNullData(scriptVersion uint16, script []byte, expectedLength int) bool {
+	// The only currently supported script version is 0.
+	if scriptVersion != 0 {
+		return false
+	}
+
+	// The script can't possibly be a null data script if it doesn't start
+	// with OP_RETURN.  Fail fast to avoid more work below.
+	if len(script) < 2 || script[0] != OP_RETURN {
+		return false
+	}
+
+	// OP_RETURN followed by data push of the expect size.
+	tokenizer := MakeScriptTokenizer(scriptVersion, script[1:])
+	return tokenizer.Next() && tokenizer.Done() &&
+		(isSmallInt(tokenizer.Opcode()) || tokenizer.Opcode() <= OP_DATA_75) &&
+		len(tokenizer.Data()) == expectedLength &&
+		expectedLength == int(tokenizer.Opcode())
```

Comment:
This is most certainly incorrect logic.  The small integer opcodes are only a single byte, so there is no way this check is correct.  I believe you want:

```Go
len(tokenizer.Data()) == expectedLength
```

Reply:
`OP_DATA_75` is true at this point and therefore we know that the opcode == the length. The test you suggest also happens one line above. This is super explicit but removing it would be fine because the len would match the opcode. This is your call if you want to remove it or not.

Reply:
Fix has been committed based on your playground code.